common/types.go

```diff
@@ -234,16 +237,302 @@ func init() {
 func BytesToAddress(b []byte) Address {
+	if IsInChainScope(b) {
+		var i InternalAddress
+		i.SetBytes(b)
+		return Address{&i}
+	} else {
+		var e ExternalAddress
+		e.SetBytes(b)
+		return Address{&e}
+	}
+}
+
+func Bytes20ToAddress(b [20]byte) Address {
+	return BytesToAddress(b[:])
+}
+
+func InternalToAddress(i *InternalAddress) Address {
 	var a Address
-	a.SetBytes(b)
+	a.inner = i
 	return a
 }
 
+// EncodeRLP serializes b into the Ethereum RLP block format.
+func (a Address) EncodeRLP(w io.Writer) error {
+	if a.inner == nil {
+		a.inner = &InternalAddress{}
+	}
+	return rlp.Encode(w, a.inner)
+}
+
+// DecodeRLP decodes the Ethereum
+func (a *Address) DecodeRLP(s *rlp.Stream) error {
+	temp := make([]byte, 0, 20)
+	if err := s.Decode(&temp); err != nil {
+		return err
+	}
+	*a = BytesToAddress(temp)
+	return nil
+}
+
+// SetBytes sets the address to the value of b.
+// If b is larger than len(a), b will be cropped from the left.
+func (a *Address) SetBytes(b []byte) {
+	if IsInChainScope(b) {
+		if internal, ok := a.inner.(*InternalAddress); ok { // Note that in the case that they do not match, nothing happens
+			internal.SetBytes(b)
+		} else if a.inner == nil {
+			var i InternalAddress
+			i.SetBytes(b)
+			a.inner = &i
+		} else {
+			fmt.Println("Tried to set an external address to an internal address")
+		}
+	} else {
+		if external, ok := a.inner.(*ExternalAddress); ok {
+			external.SetBytes(b)
+		} else if a.inner == nil {
+			var external ExternalAddress
+			external.SetBytes(b)
+			a.inner = &external
+		} else {
+			fmt.Println("Tried to set an internal address to an external address")
+		}
+	}
+}
+
+// Bytes gets the string representation of the underlying address.
+func (a Address) Bytes() []byte {
+	if a.inner == nil {
+		return []byte{}
+	}
+	return a.inner.Bytes()
+}
+
+// Hash converts an address to a hash by left-padding it with zeros.
+func (a Address) Hash() Hash {
+	if a.inner == nil {
+		return Hash{}
+	}
+	return a.inner.Hash()
+}
+
+// Hex returns an EIP55-compliant hex string representation of the address.
+func (a Address) Hex() string {
+	if a.inner == nil {
+		return string([]byte{})
+	}
+	return a.inner.Hex()
+}
+
+// String implements fmt.Stringer.
+func (a Address) String() string {
+	if a.inner == nil {
+		return string([]byte{})
+	}
+	return a.inner.String()
+}
+
+// Format implements fmt.Formatter.
+// Address supports the %v, %s, %v, %x, %X and %d format verbs.
+func (a Address) Format(s fmt.State, c rune) {
+	if a.inner != nil {
+		a.inner.Format(s, c)
+	}
+}
+
+// MarshalText returns the hex representation of a.
+func (a Address) MarshalText() ([]byte, error) {
+	if a.inner == nil {
+		return []byte{}, nil
+	}
+	return a.inner.MarshalText()
+}
+
+// UnmarshalText parses a hash in hex syntax.
+func (a *Address) UnmarshalText(input []byte) error {
+	var temp InternalAddress
+	if err := hexutil.UnmarshalFixedText("Address", input, temp[:]); err != nil {
+		return err
+	}
+	if temp.IsInChainScope() {
```

Comment:
Correct me if I am wrong, but this entire `if/else` scaffold is a duplication of the constructor logic to construct an address from a collection of bytes, right? I'd prefer to reuse that logic, rather than rewrite it. i.e. something like:
```
func (a *Address) UnmarshalText(input []byte) error {
    var temp [2]bytes
    if err := hexutil.UnmarshalFixedText("Address", input, temp[:]); err != nil {
        return err
    }
    if addr, ok := NewAddress(temp).(*InternalAddress); ok {
        *a = *addr
    } else {
        return fmt.Errorf("failed to unmarshal address from bytes %s", input)
    }
}
```

Reply:
Agreed, fixed